src/io/fastx.rs

```diff
@@ -0,0 +1,597 @@
+
+const BASE_QUAL: u8 = 30;
+
+pub enum Reader<R: io::Read> {
+    FASTA(fasta::Reader<R>),
+    FASTQ(fastq::Reader<R>),
+    Uninitialized(Option<R>),
+}
+
+pub enum Record {
+    FASTA(fasta::Record),
+    FASTQ(fastq::Record),
+}
+
+impl From<fasta::Record> for Record {
+    fn from(f: fasta::Record) -> Self {
+        Record::FASTA(f)
+    }
+}
+
+impl Into<fasta::Record> for Record {
+    fn into(self) -> fasta::Record {
+        match self {
+            Record::FASTA(f) => f,
+            Record::FASTQ(f) => fasta::Record::with_attrs(f.id(), f.desc(), f.seq()),
+        }
+    }
+}
+
+impl Into<fastq::Record> for Record {
+    fn into(self) -> fastq::Record {
+        match self {
+            Record::FASTA(f) => {
+                let q = vec![BASE_QUAL; f.seq().len()];
+                fastq::Record::with_attrs(f.id(), f.desc(), f.seq(), &q)
+            }
+            Record::FASTQ(f) => f,
+        }
+    }
+}
+
+impl From<fastq::Record> for Record {
+    fn from(f: fastq::Record) -> Self {
+        Record::FASTQ(f)
+    }
+}
+
+pub struct Records<R: io::Read> {
+    reader: Reader<R>,
+    error_has_occured: bool,
+}
+
+impl Record {
+    pub fn new() -> Self {
+        Record::FASTA(fasta::Record::new())
+    }
+
+    pub fn with_attrs(
+        id: &str,
+        desc: Option<&str>,
+        seq: TextSlice<'_>,
+        qual: Option<&[u8]>,
+    ) -> Self {
+        match qual {
+            Some(q) => Record::FASTQ(fastq::Record::with_attrs(id, desc, seq, q)),
+            None => Record::FASTA(fasta::Record::with_attrs(id, desc, seq)),
+        }
+    }
+
+    pub fn id(&self) -> &str {
+        match self {
+            Record::FASTA(x) => x.id(),
+            Record::FASTQ(x) => x.id(),
+        }
+    }
+
+    pub fn check(&self) -> Result<(), &str> {
+        match self {
+            Record::FASTA(x) => x.check(),
+            Record::FASTQ(x) => x.check(),
+        }
+    }
+
+    pub fn desc(&self) -> Option<&str> {
+        match self {
+            Record::FASTA(x) => x.desc(),
+            Record::FASTQ(x) => x.desc(),
+        }
+    }
+
+    pub fn seq(&self) -> TextSlice<'_> {
+        match self {
+            Record::FASTA(x) => x.seq(),
+            Record::FASTQ(x) => x.seq(),
+        }
+    }
+
+    pub fn qual(&self) -> Option<&[u8]> {
+        match self {
+            Record::FASTA(_) => None,
+            Record::FASTQ(x) => Some(x.qual()),
+        }
+    }
+
+    pub fn set_base_qual(&mut self, qual: &[u8]) {
+        let new_record = fastq::Record::with_attrs(self.id(), self.desc(), self.seq(), qual);
+        mem::replace(self, Record::FASTQ(new_record));
+    }
+}
+
+impl SequenceRead for Record {
+    fn name(&self) -> &[u8] {
+        self.id().as_bytes()
+    }
+
+    fn base(&self, i: usize) -> u8 {
+        self.seq()[i]
+    }
+
+    fn base_qual(&self, i: usize) -> u8 {
+        match self {
+            Record::FASTQ(x) => x.base_qual(i),
+            Record::FASTA(_) => BASE_QUAL,
```

Comment:
@HenningTimm suggestions for this case? I can't change the `base_qual` signature in `SequenceRead` trait, but panicking seems extreme...

Reply:
When using two structs with a common trait instead of an enum for the two kinds of records, you could simply implement SequenceRead only for the FASTQ case.

Reply:
I [noticed][0] the primary crate implementing `SequenceRead` is [rust-htslib][1], for BAM records (and also here for `fastq::Record`). The primary client for the `SequenceRead` trait is the [CalcConsensus][2] trait from rust-bio-tools, but they call [`qual()`][3] [directly][4], instead of `base_qual`.

But there is a wrinkle: while `MAPQ` must always be present in a SAM/BAM record, if the value is `255` it means the mapping quality is not available [SAM specs][5]. In this case, should `base_qual` be defined as `fn base_qual(&self, i: usize) -> Option<u8>` instead? Or I can change BASE_QUAL to
`const BASE_QUAL: u8 = 30;`
and match SAM/BAM expectations.

[0]: https://github.com/search?l=Rust&q=SequenceRead&type=Code
[1]: https://github.com/rust-bio/rust-htslib/blob/c32b9c4165d5e3192cfdae630a90a34e32699ffb/src/bam/record.rs#L623
[2]: https://github.com/rust-bio/rust-bio-tools/search?q=CalcConsensus&unscoped_q=CalcConsensus
[3]: https://github.com/rust-bio/rust-bio-tools/blob/8266a88c43a125ae572f19353d04456c986e8d2d/src/bam/call_consensus_reads/calc_consensus.rs#L91
[4]: https://github.com/rust-bio/rust-bio-tools/blob/8266a88c43a125ae572f19353d04456c986e8d2d/src/fastq/call_consensus_reads/calc_consensus.rs#L92
[5]: http://samtools.github.io/hts-specs/SAMv1.pdf

Reply:
about converting to a common trait instead of enum: my experience is that trait objects are harder to pass around than enums, so that's why I went this way. But I can change the implementation if people prefer it.

Reply:
base_qual and MAPQ are two different things though. MAPQ is not handled by the SequenceRead trait at all. Or am I misunderstanding you?

Reply:
I thnik here, to solve the base_qual issue, it would make sense.

Reply:
Did you come to a conclusion @luizirber?

Reply:
Sort of: I don't implement `SequenceRead` for `fastx::Record`. You can do `.set_base_qual(qual)` to have a `fastx::Record` with `Some(qual)`, or be braver and do `.try_into()` and deal with quality not being set. This removes the need for a default `BASE_QUAL` value, still allow parsing FASTA/Q files if you don't need quality, and provides a path for dealing with cases where you want quality scores without depending on something hardcoded in rust-bio.